src/wallet/wallet.cpp

```diff
@@ -4628,36 +4651,49 @@ CKeyPool::CKeyPool(const CPubKey& vchPubKeyIn, bool internalIn)
     m_pre_split = false;
 }
 
-void CWalletTx::SetMerkleBranch(const uint256& block_hash, int posInBlock)
+void CWalletTx::SetMerkleBranch(const uint256& block_hash, int posInBlock, int block_height)
 {
     // Update the tx's hashBlock
     hashBlock = block_hash;
 
     // set the position of the transaction in the block
     nIndex = posInBlock;
+
+    // set tx height
+    m_block_height = block_height;
 }
 
-int CWalletTx::GetDepthInMainChain(interfaces::Chain::Lock& locked_chain) const
+int CWalletTx::GetDepthInMainChain() const
 {
+    if (pwallet == nullptr)
+        return 0;
+    AssertLockHeld(pwallet->cs_wallet);
+    assert(m_block_height >= -1);
     if (hashUnset())
         return 0;
 
-    return locked_chain.getBlockDepth(hashBlock) * (nIndex == -1 ? -1 : 1);
+    // If nIndex is -1 and the block hash is set, it means the transaction is conflicted, and that m_block_height
```

Comment:
I think an alternative would be to reset the `hashBlock`, `nIndex` and `m_block_height` for all transactions matching the block hash when a `BlockDisconnected` is received. That would involve iterating through all of the wallet txs to check their `hashBlock`s. Re-orgs are supposed to be pretty rare, so I think that's ok. Thoughts?

Reply:
Yes I've thought about while working on the rescan logic. I've introduced a `Rewind` callback in the `Chain::Notifications`, it should call a wallet method to zeroed out cleanly tx `hashBlock`, `nIndex` and `m_block_height`. I've proposed to introduce in this future-PR, this one is already wide enough.